subsys/settings/settings_fcb.c

```diff
@@ -0,0 +1,642 @@
+/* Copyright (c) 2019 Laczen
+ * Copyright (c) 2018 Nordic Semiconductor ASA
+ * Copyright (c) 2015 Runtime Inc
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <errno.h>
+#include <string.h>
+
+#include <settings/settings.h>
+#include "settings_priv.h"
+
+struct settings_fcb_read_fn_arg {
+	struct fcb_entry_ctx *entry_ctx;
+	off_t off;
+	u8_t rbs;
+};
+
+static int settings_fcb_load(struct settings_store *cs);
+static int settings_fcb_save(struct settings_store *cs, const char *name,
+			     const char *value, size_t val_len);
+
+static struct settings_store_itf settings_fcb_itf = {
+	.csi_load = settings_fcb_load,
+	.csi_save = settings_fcb_save,
+};
+
+int settings_fcb_dst(struct settings_fcb *cf)
+{
+	cf->cf_store.cs_itf = &settings_fcb_itf;
+	settings_dst_register(&cf->cf_store);
+
+	return 0;
+}
+
+int settings_fcb_src(struct settings_fcb *cf)
+{
+	cf->cf_store.cs_itf = &settings_fcb_itf;
+	settings_src_register(&cf->cf_store);
+
+	return 0;
+}
+
+static int settings_fcb_read(struct fcb_entry_ctx *entry_ctx, off_t off,
+			     char *buf, size_t *len)
+{
+
+	if (off > entry_ctx->loc.fe_data_len) {
+		return -EINVAL;
+	}
+
+	if ((off + *len) > entry_ctx->loc.fe_data_len) {
+		*len = entry_ctx->loc.fe_data_len - off;
+	}
+
+	return flash_area_read(entry_ctx->fap,
+			       FCB_ENTRY_FA_DATA_OFF(entry_ctx->loc) + off,
+			       buf, *len);
+}
+
+static ssize_t settings_fcb_read_fn(void *back_end, void *data, size_t len)
+{
+	int rc, rem;
+	struct settings_fcb_read_fn_arg *rd_fn_arg;
+	char temp_buf[16]; /* buffer for fit read-block-size requirements */
+	size_t len_read, tmp_len;
+	off_t off;
+	u8_t *data8 = (u8_t *)data;
+
+	len_read = 0;
+	rd_fn_arg = (struct settings_fcb_read_fn_arg *)back_end;
+	off = rd_fn_arg->off;
+
+	/* Take care of alignment */
+	rem = rd_fn_arg->off % rd_fn_arg->rbs;
+
+	if (rem) {
+		/* Read from aligned offset, keep only required data */
+		off -= rem;
+		len_read = MIN(rd_fn_arg->rbs - rem, len);
+		tmp_len = rd_fn_arg->rbs;
+		rc = settings_fcb_read(rd_fn_arg->entry_ctx, off, temp_buf,
+				       &tmp_len);
+		if (rc) {
+			return rc;
+		}
+		memcpy(data8, temp_buf+rem, len_read);
+		len -= len_read;
+		data8 += len_read;
+		off += rd_fn_arg->rbs;
+	}
+
+	/* If there is more data read it */
+	if (len) {
+		tmp_len = len;
+		rc = settings_fcb_read(rd_fn_arg->entry_ctx, off, data8,
```

Comment:
Data can have end-up at unaligned position. 
Fcb data field always begin at aligned address (how ever this function can be used for readout from the middle of it, which is processed properly a few lines above), It also physically ends at aligned bloc end, but some very last bytes might be just the filler.
```
aligned ----------------------
 |                            |
\/                            \/
 | payload | filler <0,wbs-1> |
      ^    ^
      |    |
      |    payload data end might be unaligned
      |    
      - redout might be requested from unaligned location
        as well
```

Reply:
This is exactly what the routine does, it starts the read from the previous alignment, reads up to the next, if all data is inside the alignment the routine stops. if there still is data it reads until the end of the data. The routine never reads out of boundary of the fcb entry and returns the correct amount of data. Whenever a read is done outside the fcb entry an error is generated. The routine also never reads outside the payload, it shouldn't because there is no data there.

Reply:
- user might call this function with any length he want.
- fcb encodes data field len as len of payload, not payload + filler len. So that len is aligned in this read is not guaranteed.

Reply:
> * user might call this function with any length he want.
Yes, but the routine will only read/return what is inside the payload. The length that is read is returned.  
> * fcb encodes data field len as len of payload, not payload + filler len. So that len is aligned in this read is not guaranteed.
Why should len be aligned ? You can read whatever amount of bytes starting from an aligned position.

Reply:
>Why should len be aligned ? You can read whatever amount of bytes starting from an aligned position.
a flash driver  might serve only aligned access, both on write as on read. And this is true for many drivers in zephyr.

Reply:
apparently settings_fcb_read need correction as well.

Reply:
^^ looks like length read be aligned to wbs - and this is probably a bug in zephyr mainline as well :(

Reply:
I had a look at the drivers, most of them even allow unaligned reads, I don't know the drivers you are talking about. No need to make any changes to zephyr mainline, maybe a driver needs change.

Reply:
SRM drivers for example

Reply:
> SRM drivers for example

I can't find any SRM drivers in the flash directory. Maybe you mean stm32, they do not put any alignment requirements on flash reads.

Reply:
right - was STM32, indeed it have must been upgraded since I had looked at them before.
QMSI has requirement for 4 B alignment.
MCUX has not.
Geco has not.
SAM has not.
NIOS2 has requirement for 4 B alignment.
NRF has not.

Reply:
> right - was STM32, indeed it have must been upgraded since I had looked at them before.
> QMSI has requirement for 4 B alignment.
> MCUX has not.
> Geco has not.
> SAM has not.
> NIOS2 has requirement for 4 B alignment.

NIOS has no length alignment, only start

> NRF has not.

So the only one left is QMSI, that could easily be changed to remove the length alignment requirement.

Reply:
I didn't check SPI drivers -  if w25qxxdv has any requirement, it mus be depend on spec. (I didn't catch anything in its code).
And there is also spi_nor driver.

Reply:
> I didn't check SPI drivers - if w25qxxdv has any requirement, it mus be depend on spec. (I didn't catch anything in its code).

No alignment requirements

> And there is also spi_nor driver.

No alignment requirements